metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -1420,6 +1421,58 @@ class MetalsLanguageServer(
     }
   }
 
+  /**
+   * Returns the the definition location or reference locations of a symbol
+   * at a given text document position.
+   * If the symbol represents the definition itself, this method returns
+   * the reference locations, otherwise this returns definition location.
+   * https://github.com/scalameta/metals/issues/755
+   */
+  def definitionOrReferences(
+      position: TextDocumentPositionParams,
+      token: CancelToken = EmptyCancelToken
+  ): Future[DefinitionResult] = {
+    val source = position.getTextDocument.getUri.toAbsolutePath
+    if (source.toLanguage.isScala) {
+      (for {
+        doc <- semanticdbs.textDocument(source).documentIncludingStale
+        positionOccurrence = definitionProvider.positionOccurrence(
+          source,
+          position,
+          doc
+        )
+        occ <- positionOccurrence.occurrence
+      } yield occ) match {
+        case Some(occ) =>
+          if (occ.role.isDefinition) {
+            val referecneContext = new ReferenceContext(false)
+            val refParams = new ReferenceParams(referecneContext)
+            refParams.setTextDocument(position.getTextDocument())
+            refParams.setPosition(position.getPosition())
+            val result = referencesResult(refParams)
+            Future.successful(
+              DefinitionResult(
+                locations = result.locations.asJava,
+                symbol = result.symbol,
+                definition = None,
+                semanticdb = None
+              )
```

Comment:
Returns `symbol` in addition to the `locations` because `TestingServer` requires symbol information.
Wrapping those by `DefinitionResult` for structuring the return value, but what I actually want is something like [`Definitionresult | ReferencesResult`](https://dotty.epfl.ch/docs/reference/new-types/union-types.html) haha.

Reply:
That's right! But this time, let me leave the return value's type `Future[DefinitionResult]` because nesting `Either` inside `Future` makes things complicated and feels like to use something like `EitherT` ...